src/Proto3/Wire/Decode.hs

```diff
@@ -441,6 +443,26 @@ sfixed64 = runGetFixed64 getInt64le
 at :: Parser RawField a -> FieldNumber -> Parser RawMessage a
 at parser fn = Parser $ runParser parser . fromMaybe mempty . M.lookup fn
 
+-- | Try to parse different field numbers with their respective parsers.
+-- This is used to express alternative between possible fields
+--
+-- If no field number are present for the oneof, then the default
+-- behavior of the first parser is used.
+--
+-- TODO: contrary to the protobuf spec, in the case of multiple fields number
+-- matching the oneof content, the choice of field is biased
+-- to the order of the list, instead of being biased to the last field
+-- of group of field number is the oneof. This is related to the Map
+-- used for input that preserve order across multiple invocation of the same
+-- field, but not across a group of field.
+oneof :: [(FieldNumber, Parser RawField a)] -> Parser RawMessage a
```

Comment:
This function would be simpler if you implemented an `Alternative`/`MonadPlus` instance for `Parser`, because then it would just be:

```haskell
oneof kvs = msum [ at v k | (k, v) <- kvs ]
```

Reply:
yes but that would not make it closer to have the right behaviour; this still rely on a bias towards the definition order instead of the expected protobuf behaviour of the bias in favor of the last element of the set of possible/expected (by oneof) fields.